### PR TITLE
Update bindings

### PR DESCRIPTION
**Pull request by ftalbrecht**
This is the analogously branch to the `update-bindings` branch in dune-xt. Still centered around CG and DG methods for advection diffusion reaction, the bindings are already in a good state. See https://github.com/ftschindler-work/dune-gdt-python-bindings for a demo.

### CODE REVIEW

File: python/setup.py.in [hunk 1]
```diff
@@ -23,5 +23,6 @@ setup(name='dune.gdt',
       author_email='dune-gdt-dev@listserv.uni-muenster.de',
       url='https://github.com/dune-community/dune-gdt',
       packages=['dune.gdt'],
-      install_requires=['jinja2', 'where', 'codecov', 'bokeh', 'pymumps', 'metis'],
+      install_requires=['jinja2', 'where', 'codecov'],
+      extrast_requires={'parallel': ['bokeh', 'pymumps', 'metis']},
```
Comment (renefritze) [inline python/setup.py.in:27]: Got a typo here

**Comment by ftalbrecht** [comment]
merge when ready ^^

**Comment by renefritze** [comment]
> merge when ready ^^

:roll_eyes:

**Comment by renefritze** [comment]
There you go

**Comment by renefritze** [comment]
Alright, so actually this doesn't "go" right away. The action is setup to merge by rebase, which doesn't work for this PR. You'll still manually have to push the button to create a merge commit, @ftalbrecht

**Comment by ftalbrecht** [comment]
> There you go

thx, forgot we have that

**Comment by renefritze** [comment]
> thx, forgot we have that

Well, we have that now with https://github.com/dune-community/dune-gdt/pull/187/commits/10b902887df691c0593747224a13cfde7de01992

**Comment by ftalbrecht** [comment]
This is the current state which should go into master and is ready from my side, @renefritze

**Comment by ftalbrecht** [comment]
There is a [CI failure](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-gdt/-/jobs/422529/raw) associated with `EntropyBasedFluxImplementation` (search for `object of abstract class type` in the raw log), care to take a look, @tobiasleibner? I already fixed and pushed the `DiscreteValuedGridFunction` (not related)...

**Comment by tobiasleibner** [comment]
Its probably due to the function changes in ``dune-xt``, I will take a look.

**Comment by tobiasleibner** [comment]
The momentmodel tests should be fixed now.

**Comment by ftalbrecht** [comment]
The error is in dune-xt, I will restart after https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/merge_requests/46 has been merged and dune-gdt-super has been updated.

**Comment by ftalbrecht** [comment]
recent fixes require https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/merge_requests/49 to pass and be present in dune-gdt-super

**Comment by ftalbrecht** [comment]
Builds are now passing with gcc, which is a good start. The clang ones are failing with the sanitizer checks (division by zero in slope reconstructor), care to take a look, @tobiasleibner?

**Comment by tobiasleibner** [comment]
Everything seems to be fixed here, except that the ``test_burgers__1d__explicit__*`` tests are now timing out for the clang builds, probably due to the new sanitizer flags. I see three options to fix these timeouts:
- Increase the timeout limit for these (or all) tests. This is probably the easiest option but has the disadvantage that the CI runs will take longer. 
- Use `-O1` for the clang builds. We probably have to do this also for ``dune-xt`` if we do not want to change our test setup. If the numbers for ``dune-gdt`` are similar to [the numbers for ``dune-xt``](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/issues/5#note_98367), this will increase compilation time by a factor of about 3 and memory usage during compilation by a factor of 2.
-  Reduce the test size. This would be my favorite option as it reduces the CI run time. Profiling of the first 10 minutes of``test_burgers__1d__explicit__fv`` shows that all of the time is spent in the timestep estimation (``estimate_fixed_explicit_dt_to_T_end``), maybe we can just prescribe the dt for the tests (And maybe add another test that tests the timestep estimation function?). Also, a lot of time during the timestep estimation is spent in the ``WithLogger`` class, maybe we can just turn off logging for these tests? (Also, we should rename the arguments ``start_enabled`` (``WithLogger`` class) and ``start_disabled`` (``DefaultLogger`` class) to be the same (both should use ``start_enabled`` or both ``start_disabled``), otherwise this is quite confusing.

**Comment by ftalbrecht** [comment]
> Everything seems to be fixed here, except that the `test_burgers__1d__explicit__*` tests are now timing out for the clang builds, probably due to the new sanitizer flags. I see three options to fix these timeouts:
> 
>     * Increase the timeout limit for these (or all) tests. This is probably the easiest option but has the disadvantage that the CI runs will take longer.
> 
>     * Use `-O1` for the clang builds. We probably have to do this also for `dune-xt` if we do not want to change our test setup. If the numbers for `dune-gdt` are similar to [the numbers for `dune-xt`](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/issues/5#note_98367), this will increase compilation time by a factor of about 3 and memory usage during compilation by a factor of 2.
> 
>     * Reduce the test size. This would be my favorite option as it reduces the CI run time. Profiling of the first 10 minutes of`test_burgers__1d__explicit__fv` shows that all of the time is spent in the timestep estimation (`estimate_fixed_explicit_dt_to_T_end`), maybe we can just prescribe the dt for the tests (And maybe add another test that tests the timestep estimation function?). Also, a lot of time during the timestep estimation is spent in the `WithLogger` class, maybe we can just turn off logging for these tests? (Also, we should rename the arguments `start_enabled` (`WithLogger` class) and `start_disabled` (`DefaultLogger` class) to be the same (both should use `start_enabled` or both `start_disabled`), otherwise this is quite confusing.

I agree with the latter, taking a look right now. Its not that straight forward however: we can easily have different dts for different tests using the config setup/expectation system that we have right now. However, each run consists of several solves (several refinements and reference solution) which all use the same generic solve() method (which is quite nice btw). Thus, we cannot determine within the solve which refinement we are on and thus cannot use different dts for different grids.

My quick solution is to use the smallest dt everywhere, which is far from optimal in a mathematical/document what should be done sense. However, using the .mini files is optional, and when executing the executables directly without ctest, one still gets the desired behavior, so this is all-in-all fine with me. Will push a fix soon, if no one objects...

**Comment by tobiasleibner** [comment]
The CI runs now pass, except for ``unstable_gcc``, where one tests fails. We probably have to adjust the tolerance, @ftalbrecht, then this PR should be good to go.

**Comment by ftalbrecht** [comment]
On 13.10.20 14:22, Tobias Leibner wrote:
> The CI runs now pass, except for ``unstable_gcc``, where one tests fails. We probably have to adjust the tolerance, @ftalbrecht, then this PR should be good to go.
> 

I'll look into it

**Comment by ftalbrecht** [comment]
This is one of those cases where we get a different close-to-zero from a different compiler. Not sure yet how to handle these nicely.

**Comment by ftalbrecht** [comment]
Once https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/merge_requests/54 is through, we have the means to be more flexible with the zero detection in the tests.

**Comment by ftalbrecht** [comment]
- [ ] Dockerfiles reverted
- [ ] dune-xt-super images updated
- [ ] dune-xt pr merged
- [ ] dune-gdt-super images updated
- [ ] this pr restarted

**Comment by ftalbrecht** [comment]
I pushed another fix to dune-xt, same as above:
- [ ] [dune-xt pr](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/merge_requests/60): [pipeline status](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-xt/-/commits/another-try-to-fix-zero-tolerance-comparison)
- [ ] dune-gdt-super: commit updated dune-xt master
- [ ] dune-gdt-super: updated test images
- [ ] this pr restarted

**Comment by tobiasleibner** [comment]
For the ``clang`` checks to work, we have to wait for the [updated docker images](https://github.com/dune-community/Dockerfiles/pull/95) and then update the ``dune-gdt-super`` module again. If we do not want to wait for this, a quick fix would be to  
```bash
export ASAN_SYMBOLIZER_PATH=/usr/bin/llvm-symbolizer-${CLANG_VERSION}
```
 in the CI scripts (e.g., in ``.ci/shared/test_cpp.bash``).

**Comment by tobiasleibner** [comment]
The ``clang`` builds are fixed, if you fix the remaining test failure, this PR should be good to go, @ftalbrecht.

**Comment by ftalbrecht** [comment]
> The `clang` builds are fixed, if you fix the remaining test failure, this PR should be good to go, @ftalbrecht.

Ok it. Though I see all builds [failed here](https://zivgitlab.uni-muenster.de/ag-ohlberger/dune-community/dune-gdt/-/commit/dfa4c9e60e4dad576befce4322ad33f92aae84a0/pipelines?ref=update-bindings), or am I looking in the wrong place?

**Comment by tobiasleibner** [comment]
No, you are right, "fixed" was a little misleading. All three jobs are failing due to the same test failure now. Before, every single test in the ``clang`` pipeline was failing due to an address sanitizer error.

**Comment by ftalbrecht** [comment]
Great work!

**Pull request merged**